src/sqlfluff/rules/L064.py

```diff
@@ -0,0 +1,262 @@
+       support single *and* double quotes automated fixes can potentially break working
+       SQL code.
+
+       This rule is disabled by default for dialetcs that don't allow double quotes for
+       STRING datatype (e.g. Postgres). It can be enabled with the
+       ``force_enable = True`` flag.
+
+    **Anti-pattern**
+
+    .. code-block:: sql
+       :force:
+
+        select
+            "abc",
+            'abc',
+            "\"",
+            "abc" = 'abc'
+        from foo
+
+    **Best practice**
+
+    Ensure all STRING literals use preferred quotes, unless escaping can be reduced by
+    using alternate quotes.
+
+    .. code-block:: sql
+       :force:
+
+        select
+            "abc",
+            "abc",
+            '"',
+            "abc" = "abc"
+        from foo
+
+    """
+
+    config_keywords = ["preferred_string_quotes", "force_enable"]
+    _dialects_with_double_quoted_strings = [
+        "bigquery",
+        "hive",
+        "mysql",
+        "sparksql",
+    ]
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        # Config type hints
+        self.preferred_string_quotes: str
+        self.force_enable: bool
+
+        if (
+            context.dialect.name not in self._dialects_with_double_quoted_strings
+            and not self.force_enable
+        ):
+            return LintResult(memory=context.memory)
+
+        # Only care about quoted literal segments.
+        if not context.segment.name == "quoted_literal":
+            return None
+
+        # If quoting style is set to consistent we use the quoting style of the first
+        # quoted_literal that we encounter.
+        if self.preferred_string_quotes == "consistent":
+            memory = context.memory
+            preferred_string_quotes = memory.get("preferred_string_quotes")
+
+            if not preferred_string_quotes:
+                # Getting the quote from LAST character to be able to handle STRING
+                # prefixes
+                preferred_string_quotes = (
+                    "double_quotes"
+                    if context.segment.raw[-1] == '"'
+                    else "single_quotes"
+                )
+                memory["preferred_string_quotes"] = preferred_string_quotes
+                self.logger.debug(
+                    "Preferred string quotes is set to `consistent`. Derived quoting "
+                    "style %s from first quoted literal.",
+                    preferred_string_quotes,
+                )
+        else:
+            preferred_string_quotes = self.preferred_string_quotes
+
+        fixed_string = self._normalize_preferred_string_quotes(
+            context.segment.raw,
+            preferred_quote_char=QUOTES_MAPPING[preferred_string_quotes][
+                "preferred_quote_char"
+            ],
+            alternate_quote_char=QUOTES_MAPPING[preferred_string_quotes][
+                "alternate_quote_char"
+            ],
+        )
```

Comment:
Slight performance nit, that you're calling this even for the first iteration (when preferred_quote is not set so will be whatever it is currently, so no need to fix).

Should you exit early above?

Or do you still need this for the case when the first quote contains an unnecessary escaped quote?

Reply:
Yeah, exactly. Just because the quotes match already, doesn't mean we are not going to touch the string. We will go into this case https://github.com/sqlfluff/sqlfluff/pull/3118/files#diff-c74eb3c9d616e01ddba16d42924a3a06bee0828c413408b642f6255e8c3a53d7R186-R190